Re-organize contrib docs, add sphinxext-rediraffe, and Dockerize (#296) (#307)
rockstor/rockstor-doc: contribute/contribute_documentation.rst

        [you@your_laptop rockstor-doc]# git checkout -b issue#1234_brief_label

You can then start making changes in this branch.

Guidelines
^^^^^^^^^^

To keep in line with Rockstor's goal to make its features as accessible as
possible, this documentation should strive to keep non-technical users as its
primary target. As such, the use of external references and links to additional
documentation to provide the reader with further technical information is
encouraged.

> If you are generating videos, and prefer uploading your videos to the Rockstor
> Youtube channel, please send an email to support@rockstor.com.

Reviewer: ```suggestion
```
Given we have recently lost the prior main Rockstor YouTube channel of Rockstorinc and have only the barely visible/populated "Rockstor" channel left, and not under my direct control (we have to bother Suman for that) I'm thinking of cutting our loses and keeping things as simple as possible and removing, in time, all YouTube links. I'm not myself keen on the video links and they are a lot of work to maintain. Once we have openQA testing in place we may have a source to reproduce current functions and even, potentially, video generation on-hand. But we are not there yet unfortunately. Hence suggesting we scrub this suggestion as we already have a backlog of now 'lost' videos that we either have to re-create and host or work around them.

If you use our :ref:`docker image<dockersphinx>`, you can use the following
command:

.. code-block:: bash

        [you@your_laptop /path/to/rockstor-doc]# docker run --rm -v $PWD:/docs ghcr.io/rockstor/rockstor-doc:main make html

HTML files are generated in the :code:`_build/html` directory. From a separate
terminal window, you can have a simple Python webserver always serving up this
content with the following command:

.. code-block:: bash

        [you@your_laptop /path/to/rockstor-doc/_build/html]# python -m http.server 8000

You can now go to :code:`http://localhost:8000` in your browser to review your
changes. The webserver is to be started only once and it will continue to serve
the files and changes you make to them.

After making any changes to a *.rst* file, run :code:`make html` as shown above
and refresh your browser to display your changes.

Submit your changes
-------------------

Once you are satisfied with your changes, you can start preparing them for
submission.

Add and commit your changes
^^^^^^^^^^^^^^^^^^^^^^^^^^^

First, let's add your changes:

.. code-block:: bash

        [you@your_laptop rockstor-doc]# git add new_file_added.rst existing_file.rst

> Then, you can commit them. We strongly encourage you to commit changes a
> certain way to help other contributors and keep the merge process smooth. The

Reviewer: ```suggestion
Then, you can commit them. We strongly encourage you to commit changes in a
certain way to help other contributors, and to keep the merge process smooth. The
```
Trivial readability suggestion.

<https://github.com/rockstor/rockstor-doc/blob/master/AUTHORS>`_ file.

Moving pages
^^^^^^^^^^^^

If your changes involve a page relocation or removal, we need to ensure any
eventual external link to it remains valid and provide a valid redirection. To
do so, we leverage the excellent Sphinx extension `sphinxext-rediraffe <https://github.com/wpilibsuite/sphinxext-rediraffe>`_.
Indeed, Rediraffe can simplify the process by comparing your current Git branch
to your *master* branch and automatically write redirections for pages that
were renamed or relocated.
To do so, you simply need to run the :code:`rediraffewritediff` builder:

.. code-block:: bash

        [you@your_laptop /path/to/rockstor-doc]# sphinx-build -b rediraffewritediff . _build/rediraffe

> If you use Docker image, you can use the following command:

Reviewer: ```suggestion
If you use the Docker image, you must use the following command:
```
Minor word change suggestion.

.. code-block:: bash

        [you@your_laptop /path/to/rockstor-doc]# docker run --rm -v $PWD:/docs ghcr.io/rockstor/rockstor-doc:main sphinx-build -b rediraffewritediff . _build/rediraffe

You should now see the needed redirects in :code:`redirects.txt`.

.. note::
        Make sure to commit your changes with Git **before** running the :code:`rediraffewritediff`

>         builder as the latter will otherwise not be able to detect your changes.

Reviewer: Nice addition.

> While we strive to limit such occasion, special circumstances might require the
> deletion of page. As **sphinxext-rediraffe** cannot yet automatically write a

Reviewer: ```suggestion
While we strive to limit such occasions, special circumstances might require the
deletion of one or more pages. As **sphinxext-rediraffe** cannot yet automatically write a
```
Minor re-wording suggestion.

redirection for a deleted page, one needs to manually instruct it. Fortunately,

> this is a simple as writing a new line in :code:`redirects.txt`, listing the

Reviewer: ```suggestion
this is as simple as writing a new line in :code:`redirects.txt`, listing the
```
typo
